### PR TITLE
FB8-131: Option to shutdown master if semi-sync is being switched off while there are active un-acked trxns

### PR DESCRIPTION
**Pull request by inikep**
Jira issue: https://jira.percona.com/browse/FB8-131

Reference Patch: https://github.com/facebook/mysql-5.6/commit/2d47615
Reference Patch: https://github.com/facebook/mysql-5.6/commit/d1ea741
Reference Patch: https://github.com/facebook/mysql-5.6/commit/f5085f4
Reference Patch: https://github.com/facebook/mysql-5.6/commit/9204c65

---------- https://github.com/facebook/mysql-5.6/commit/2d47615 ----------

Option to shutdown master if semi-sync is being switched off while there are active un-acked trxns

Summary:
Semi-sync master can be switched off manually or because of a timeout while
waiting for an ack.

This change introduces a new variable rpl_semi_sync_master_crash_if_active_trxs
which when set to true will cause the master to shutdown (after printing an
error msg) if semi-sync is being switched off while there are active un-acked
transactions. This prevents any un-acked trx from committing on the
master.

Originally Reviewed By: hermanlee

fbshipit-source-id: https://github.com/facebook/mysql-5.6/commit/c378441cb197f3a2759b8a81814ec0905caf0ed8

---------- https://github.com/facebook/mysql-5.6/commit/d1ea741 ----------

Fix test flakiness in rpl.rpl_semi_sync-master_crash_if_active_trxs

Summary:
The test would sometimes fail if the slave's retry count is exhausted
before the master comes up. Set the slave to retry forever instead.

Squash with D5841339

Originally Reviewed By: abhinav04sharma

fbshipit-source-id: https://github.com/facebook/mysql-5.6/commit/c12f822a4df784d95fb99b308d03c8feb4309327

---------- https://github.com/facebook/mysql-5.6/commit/f5085f4 ----------

Use master_retry_count to fix rpl.rpl_semi_sync_master_crash_if_active_trxs

Summary:
The proper variable to adjust is master_retry_count, not
master_connect_retry.

Squash with D5841339

Originally Reviewed By: abhinav04sharma

fbshipit-source-id: https://github.com/facebook/mysql-5.6/commit/6938715f85058fec3b6b41ddb95ac800e9ff925c

---------- https://github.com/facebook/mysql-5.6/commit/9204c65 ----------

Delete pid file when crashing after timeout with unacked trx

Summary:
If rpl_semi_sync_master_crash_if_active_trxs = true, the server crashes after a
timeout if there are any active unacked transactions. mysqld_safe expects the
pid file to be deleted if the expectation is to not bring the intance back up
again. This change deletes the pid file before crashing on this condition.

Squash with D5841339

Originally Reviewed By: Tema

fbshipit-source-id: https://github.com/facebook/mysql-5.6/commit/a94df242f3950393b257060cf15ea3f2dd651152

### CODE REVIEW

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 1]
```diff
@@ -0,0 +1,117 @@
+source include/master-slave.inc;
+source include/have_debug.inc;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test:1]: Swap

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+call mtr.add_suppression("Timeout waiting for reply of binlog");
+call mtr.add_suppression("Force shutdown: Semi-sync master is being switched off while there are active un-acked transactions");
+call mtr.add_suppression("A message intended for a client cannot be sent there as no client-session is attached");
+
+connection master;
+create table t1(a int) engine=innodb;
+--source include/sync_slave_sql_with_master.inc
+
+# Disable ack on slave
+connection slave;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]: Redundant

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+connection master;
+--source include/sync_slave_sql_with_master.inc
+
+##
+## Turning off semi-sync when there are active un-acked trxs should shutdown the
+## master
+##
+
+# Disable ack on the slave
+connection slave;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]: Redundant

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+## Turning off semi-sync when there are no active un-acked trxs should succeed
+##
+connection master;
+set @@global.rpl_semi_sync_master_enabled= OFF; # should not fail
+set @@global.rpl_semi_sync_master_enabled= ON;
+
+# Cleanup
+connection master;
+drop table t1;
+--source include/sync_slave_sql_with_master.inc
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]: Redundant

File: plugin/semisync/semisync_master_plugin.cc [hunk 5]
```diff
@@ -231,6 +231,14 @@ static MYSQL_SYSVAR_ULONG(
     fix_rpl_semi_sync_master_timeout,  // update
     10000, 0, ~0UL, 1);
 
+static MYSQL_SYSVAR_BOOL(crash_if_active_trxs,
+                         rpl_semi_sync_master_crash_if_active_trxs,
+                         PLUGIN_VAR_OPCMDARG,
+                         "Crash if there is an attempt to switch off semi-sync "
+                         "master while there are "
+                         "active un-acked transactions",
+                         NULL, NULL, 0);
```
Comment (laurynas-biveinis) [inline plugin/semisync/semisync_master_plugin.cc]: nullptr, nullptr, false?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

**Comment by inikep** [comment]
PR was updated according to comments

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+# Disable ack on slave
+source include/stop_slave.inc;
+SET GLOBAL DEBUG='+d,dont_send_semi_sync_reply';
+CHANGE MASTER TO MASTER_RETRY_COUNT=0;
+source include/start_slave.inc;
+
+##
+## Ack timeout when there are active un-acked trxs should shutdown the master
+##
+connection master;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]:
```suggestion
--source include/rpl_connection_master.inc
```

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 7]
```diff
@@ -0,0 +1,114 @@
+source include/have_debug.inc;
+source include/master-slave.inc;
+
+
+call mtr.add_suppression("Timeout waiting for reply of binlog");
+call mtr.add_suppression("Force shutdown: Semi-sync master is being switched off while there are active un-acked transactions");
+call mtr.add_suppression("A message intended for a client cannot be sent there as no client-session is attached");
+
+connection master;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]:
```suggestion
--source include/rpl_connection_master.inc
```

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+exec echo "restart" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect;
+# The master will die while executing this stmt because slave will not ack it
+error CR_SERVER_LOST;
+insert into t1 values(1);
+
+# Wait for master to die
+source include/wait_until_disconnected.inc;
+
+# Resume normal slave operations i.e start acking
+connection slave;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]:
```suggestion
--source include/rpl_connection_slave.inc
```

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+SET GLOBAL DEBUG='-d,dont_send_semi_sync_reply';
+
+# Restart master
+echo # Restart the master server;
+let $rpl_server_number= 1;
+source include/rpl_start_server.inc;
+disable_reconnect;
+
+connection master;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]:
```suggestion
--source include/rpl_connection_master.inc
```

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+## Turning off semi-sync when there are active un-acked trxs should shutdown the
+## master
+##
+
+# Disable ack on slave
+source include/stop_slave.inc;
+SET GLOBAL DEBUG='+d,dont_send_semi_sync_reply';
+source include/start_slave.inc;
+
+connection master;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]:
```suggestion
--source include/rpl_connection_master.inc
```

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+# Set timeout to a large value so that we have a chance to switch semi-sync off
+# manually
+set @@global.rpl_semi_sync_master_timeout= 1000000;
+send insert into t1 values(2); # this stmt will hang
+
+# Wait till the insert is processed on the slave
+connection slave;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]:
```suggestion
--source include/rpl_connection_slave.inc
```

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+let $wait_condition= SELECT count(*) = 1 from t1 where a = 2;
+source include/wait_condition.inc;
+
+connection master1;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]:
```suggestion
--source include/rpl_connection_master1.inc
```

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+enable_reconnect;
+exec echo "restart" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect;
+error CR_SERVER_LOST;
+set @@global.rpl_semi_sync_master_enabled= OFF; # master will shutdown
+
+# Wait for master to die
+source include/wait_until_disconnected.inc;
+
+connection master;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]:
```suggestion
--source include/rpl_connection_master.inc
```

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+error CR_SERVER_LOST;
+reap;
+
+# Resume normal slave operations i.e start acking
+connection slave;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]:
```suggestion
--source include/rpl_connection_slave.inc
```

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+SET GLOBAL DEBUG='-d,dont_send_semi_sync_reply';
+
+# Restart master
+echo # Restart the master server;
+let $rpl_server_number= 1;
+source include/rpl_start_server.inc;
+disable_reconnect;
+
+connection master;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]:
```suggestion
--source include/rpl_connection_master.inc
```

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+--source include/sync_slave_sql_with_master.inc
+
+# Will contain 2 rows because master recovery rolls trxs forward
+echo "Table at the master";
+connection master;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]:
```suggestion
--source include/rpl_connection_master.inc
```

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+select * from t1;
+
+# Will contain 2 rows because slaves received all trxs
+echo "Table at the slave";
+connection slave;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]:
```suggestion
--source include/rpl_connection_slave.inc
```

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+select * from t1;
+
+##
+## Turning off semi-sync when there are no active un-acked trxs should succeed
+##
+connection master;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]:
```suggestion
--source include/rpl_connection_master.inc
```

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+set @@global.rpl_semi_sync_master_enabled= OFF; # should not fail
+set @@global.rpl_semi_sync_master_enabled= ON;
+
+# Cleanup
+connection master;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]:
```suggestion
--source include/rpl_connection_master.inc
```

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

**Comment by inikep** [comment]
PR was updated

File: mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test (hunk 12 above)
Comment (percona-ysorokin) [reply mysql-test/suite/rpl/t/rpl_semi_sync_master_crash_if_active_trxs.test]: Not fixed

**Review by percona-ysorokin — APPROVED** [review]
LGTM with `connection master1` addressed

**Review by lth — APPROVED** [review]

**Pull request closed without merging**